### PR TITLE
feat(moe): add more activation types to the SM90 CuTe-DSL MoE backend

### PR DESCRIPTION
**Pull request by Aneureka**
## 📌 Description

Adds more activation types to the SM90 CuTe-DSL BF16/FP16 MoE backend:
- gated SwiGLU (default), including the parameterized form via `swiglu_alpha` / `swiglu_beta` / `swiglu_limit` (gate clamped from above, up clamped to `[-limit, limit]`, `gate * sigmoid(alpha * gate) * (up + beta)`) and SiTU via `situ_beta` / `situ_linear_beta`
- gated GeGLU-tanh (`ActivationType.GegluTanh`)
- non-gated ReLU² (`ActivationType.Relu2`; `w1` is a plain `[E, I, hidden]` projection, `I % 64 == 0`)

## 🔍 Related Issues

Follow-up to #4878.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Summary by CodeRabbit

* **New Features**
  * Added support for SwiGLU, SiTU, GeGLU-tanh, and non-gated ReLU² activations in BF16/FP16 fused MoE workflows.
  * Added configurable activation parameters and gated or non-gated weight layouts.
  * Added activation validation and expanded Hopper benchmark coverage across additional model shapes.

* **Documentation**
  * Documented the BF16 wrapper, activation options, supported shapes, tracing support, and benchmark coverage.

### CODE REVIEW

**Comment by Aneureka** [comment]
@flashinfer-bot run

**Comment by Aneureka** [comment]
@flashinfer-bot run

**Comment by Aneureka** [comment]
@flashinfer-bot run

File: flashinfer/fused_moe/cute_dsl/common/kernel_utils.py [hunk 1]
```diff
@@ -174,6 +175,90 @@ def silu_f32(
...
+def f32_reciprocal(value: float) -> float:
+    """Return ``1 / fp32(value)``, rounded to fp32.
+
+    The reciprocal must come from the same fp32 value the kernel multiplies back
+    in.  Taking it from the unrounded Python float instead can differ by 1 ulp for
+    betas that are not fp32-exact.  (f64 carries at least 2p+2 bits for p=24, so
+    computing the quotient in f64 and rounding once to fp32 is correctly rounded.)
+    """
+    beta_f32 = ctypes.c_float(float(value)).value
+    return ctypes.c_float(1.0 / beta_f32).value
```
[the remark this answers is not in the document]
Comment (Aneureka) [reply flashinfer/fused_moe/cute_dsl/common/kernel_utils.py:223]: Added the checks.

**Comment by Aneureka** [comment]
@flashinfer-bot run

File: flashinfer/fused_moe/cute_dsl/moe_utils.py [hunk 2]
```diff
@@ -34,6 +34,19 @@ def _get_cuda_stream_ptr() -> int:
...
+def _is_finite_fp32(value: float, *, positive: bool = False) -> bool:
+    """Whether ``value`` is finite (and positive, if requested) after fp32 rounding.
+
+    The kernels consume these constants (and, for the SiTU scales, their
+    reciprocals) as fp32, so a Python float that rounds to ``0.0`` or ``inf``
+    in fp32 is unusable even though it is finite and positive in f64.
+    """
+    if not math.isfinite(value) or (positive and value <= 0):
+        return False
+    value_f32 = torch.tensor(value, dtype=torch.float32).item()
```
Comment (feih-nv) [inline flashinfer/fused_moe/cute_dsl/moe_utils.py]: **Should fix.** `torch.tensor(value, dtype=torch.float32).item()` is a device round-trip when the default device is CUDA, and breaks graph capture: with `torch.set_default_device('cuda')` this raises `cudaErrorStreamCaptureInvalidated` inside `torch.cuda.graph()` (reproduced on B200). `validate_cute_dsl_moe_swiglu_config` runs unconditionally, so every SM90 eager forward pays it three times (op entry, `CuteDslFusedMoESm90Runner.__init__`, the GEMM1 wrapper) = 9 round-trips; the shared `validate_cute_dsl_moe_situ_config` now inherits it for SM100 too. `ctypes.c_float(value).value` (already used by `f32_reciprocal` in this PR) is a drop-in and matches the new tests (1e39 -> inf, 1e-50 -> 0.0, NaN -> NaN).

File: benchmarks/routines/moe.py [hunk 3]
```diff
@@ -1651,11 +1651,13 @@ def testCuteDslBf16Moe(args):
...
     This test:
-    1. Creates bf16/fp16 weights ([gate; up] w13 repacked to the kernel's
-       32-column up/gate interleave) and pre-routed top-k ids/scales
+    1. Creates bf16/fp16 weights (gated activations: [gate; up] w13 repacked
+       to the kernel's 32-column up/gate interleave; ReLU2: a single [E, I, H]
+       projection) and pre-routed top-k ids/scales
     2. Runs MoE via CuteDslBf16MoEWrapper (or cute_dsl_fused_moe_bf16 when
-       ``--use_functional_api`` is set). SwiGLU only. ``--autotune`` runs the
-       AutoTuner pass, compiling each candidate specialization as it is profiled.
+       ``--use_functional_api`` is set). ``--activation_type`` selects Swiglu
```
Comment (feih-nv) [inline benchmarks/routines/moe.py]: **Nit.** Docstring says `--activation_type`; argparse registers `--activation-type` (line 218, and `sample_testlist.txt` uses that spelling). argparse does not alias underscore/hyphen, so copying this gives `unrecognized arguments`. The neighbouring flags really are underscore-spelled, which is why it is easy to miss.

**Review by feih-nv — COMMENTED** [review]
**Should fix. Not regenerated: `tests/trace/fi_trace_out/cute_dsl_fused_moe_bf16_h2048_e128_topk8.json`.** `cute_dsl_fused_moe_bf16_trace` now says "SwiGLU/OAI/SiTU, GeGLU-tanh, or non-gated ReLU^2" and declares six new optional Scalars, but the committed JSON still reads "GEMM1 with SiLU-gating epilogue", `gemm1_out_size` still says "(gated SwiGLU)", and `inputs` has no `activation_type` / `swiglu_*` / `situ_*`. `fi_trace.py` emits every `spec.inputs` entry, so regenerating changes the file (CLAUDE.md Trace Template Checklist steps 4 and 7). No golden-vs-template test covers this template, so CI stays green; needs one run of `tests/trace/example.py` on SM90.

**Nit (docs). `hopper/contiguous_gather_grouped_gemm_act_fusion.py:363-365`.** `__call__` docstring still describes B as "up/gate interleaved at 32 columns" and C as `(permuted_m, N/2, 1)`; wrong for Relu2 (plain `[I, K]`, N output columns). The class, `wrapper()` and host-wrapper docstrings were updated; this one was missed.

**Follow-up, not for this PR. Gate/up source order.** `interleave_up_gate_sm90(w_gate_up)` takes `[gate; up]`, the opposite of the unified `prepare_weights` canonical `[up, gate]` (see #5291). Pre-existing, internally consistent here (`ref_activation` swaps halves explicitly), and SM90 CuTe-DSL is not in the unified API yet, so nothing is wrong today. When SM90 is wired into `MoELayer`, `prepare_weights` must take `[up, gate]` and this helper becomes a footgun. Cleanest fix later: make `_interleave_gated_halves(up, gate)` the public reference (halves as explicit args, no concatenation-order assumption) and drop the gate-first wrapper; tests then generate `[up, gate]` and the swap in `ref_activation` goes away. Until then, worth one docstring line on the helper saying it is the reverse of the unified canonical order.

File: tests/moe/test_cute_dsl_bf16_moe.py [hunk 4]
```diff
@@ -149,23 +149,37 @@ def make_random_topk(num_experts, num_tokens, top_k, device="cuda"):
...
+    from flashinfer.fused_moe import SwiGLU
+
+    activation = activation or SwiGLU()
+    if activation.is_gated:
+        inter = w1_e.shape[0] // 2
+        values = torch.cat((xe @ w1_e[inter:].T, xe @ w1_e[:inter].T), dim=-1)
+    else:
+        inter = w1_e.shape[0]
+        values = xe @ w1_e.T
+    return compute_reference_activation(values, activation, inter).float()
```
Comment (feih-nv) [inline tests/moe/test_cute_dsl_bf16_moe.py]: **Nit.** `compute_reference_activation` ends with `.to(torch.bfloat16)` (tests/moe/utils.py:147), so the pre-existing fp16 test now compares an fp16 kernel against a bf16-rounded reference; the deleted `ref_moe` was pure fp32. Tolerance (atol 3e-1) almost certainly still holds, but the reference got coarser. Rounding to `values.dtype` in `compute_reference_activation`, or `.float()` before the cast on this path, keeps the old fidelity.

File: flashinfer/fused_moe/cute_dsl/sm90_fused_moe.py [hunk 5]
```diff
@@ -232,30 +273,38 @@ def cute_dsl_fused_moe_bf16(
...
+    ``out[t] = sum_k scale[t,k] * ffn_expert(x[t]; e[t,k])`` with, for the
+    default SwiGLU,
     ``ffn(x; e) = (silu(x @ w1_gate[e].T) * (x @ w1_up[e].T)) @ w2_weight[e].T``.
 
     Supported configuration:
         * Arch: SM90 (Hopper) only.
         * Dtypes: bf16 or fp16 activations and weights (must match), fp32
           accumulation; output dtype = input dtype. No quantized paths.
-        * Activation: SwiGLU (SiLU-gated) only, fused into GEMM1.
+        * Activation: SwiGLU (SiLU-gated) etc, fused into GEMM1.
```
Comment (feih-nv) [inline flashinfer/fused_moe/cute_dsl/sm90_fused_moe.py]: **Nit (docs).** "SwiGLU (SiLU-gated) etc" replaces the exact supported set with a placeholder, and this docstring is rendered by autofunction in `docs/api/fused_moe.rst`. The wrapper docstring (486-488) and the design doc enumerate Swiglu (incl. OAI/SiTU), GegluTanh, Relu2; list them here too.

Comment (Aneureka) [reply flashinfer/fused_moe/cute_dsl/sm90_fused_moe.py]: Updated.

File: tests/moe/test_cute_dsl_bf16_moe.py (hunk 4 above)
Comment (Aneureka) [reply tests/moe/test_cute_dsl_bf16_moe.py]: Updated.

File: benchmarks/routines/moe.py (hunk 3 above)
Comment (Aneureka) [reply benchmarks/routines/moe.py]: Fixed.

File: flashinfer/fused_moe/cute_dsl/moe_utils.py (hunk 2 above)
Comment (Aneureka) [reply flashinfer/fused_moe/cute_dsl/moe_utils.py]: Good catch. Fixed.

**Comment by Aneureka** [comment]
@flashinfer-bot run

**Review by feih-nv — APPROVED** [review]
LGTM

**Comment by Aneureka** [comment]
@flashinfer-bot run
